virtool/jobs/data.py

```diff
@@ -472,3 +481,50 @@ async def force_delete(self):
         job_ids = await self._db.jobs.distinct("_id")
         await gather(*[self._client.cancel(job_id) for job_id in job_ids])
         await self._db.jobs.delete_many({"_id": {"$in": job_ids}})
+
+    async def timeout(self):
+        """
+        Timeout dead jobs.
+
+        Times out all jobs that have been running or preparing for more than 30 days.
+        This conservatively cleans up legacy jobs that do not have a ping field.
+
+        Times out all jobs that have a ping field and haven't received a ping in 5
+        minutes.
+
+        """
+        now = arrow.utcnow()
+
+        async with self._db.create_session() as session:
+            async for document in self._db.jobs.find(
+                {
+                    "state": {"$in": ["preparing", "running"]},
```

Comment:
Should waiting be on this list? If a job has been waiting for greater than 30 days, I would assume that redis state got wiped before the job could ever be ingested. Might cause problems if the `ping` field is defined on row creation though

... Also despite the syntactic verbosity, should probably use the Enum here as well

Reply:
> Should waiting be on this list? I would assume that redis state got wiped before the job could ever be ingested.

No. If Redis state somehow got wiped in production (should never happen), I would prefer to see obvious buggy behaviour in the app than hiding a serious issue behind timeout states.

If jobs somehow get stuck waiting, it would be better to just push their ids back into the list. Luckily this is a non-issue outside of flaky dev environments. I have never seen it in cloud.

> should probably use the Enum here as well

Yes